service/health/searcher/service.go

```diff
@@ -51,13 +53,84 @@ func New(config Config) (*Service, error) {
 	return s, nil
 }
 
+func extractStatusCluster(status v1alpha1.StatusCluster) ClusterInfo {
+	nodes := []NodeInfo{}
+	for _, node := range status.Nodes {
+		nodes = append(nodes, NodeInfo{
+			labels: node.Labels,
+		})
+	}
+	return ClusterInfo{
+		nodes,
+	}
+}
+
+// searchAWSCR searches for the cluster config in AWSClusterConfigs resources.
+func (s *Service) searchAWSInfo(ctx context.Context, clusterID string) (*ClusterInfo, error) {
+	cr, err := s.g8sClient.ProviderV1alpha1().AWSConfigs("default").Get(clusterID, v1.GetOptions{})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterInfo := extractStatusCluster(cr.Status.Cluster)
+	return &clusterInfo, nil
+}
+
+// searchAzureCR searches for the cluster config in AWSClusterConfigs resources.
+func (s *Service) searchAzureInfo(ctx context.Context, clusterID string) (*ClusterInfo, error) {
+	cr, err := s.g8sClient.ProviderV1alpha1().AzureConfigs("default").Get(clusterID, v1.GetOptions{})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterInfo := extractStatusCluster(cr.Status.Cluster)
+	return &clusterInfo, nil
+}
+
+// searchKVMCR abc
+func (s *Service) searchKVMInfo(ctx context.Context, clusterID string) (*ClusterInfo, error) {
+	cr, err := s.g8sClient.ProviderV1alpha1().KVMConfigs("default").Get(clusterID, v1.GetOptions{})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	clusterInfo := extractStatusCluster(cr.Status.Cluster)
+	return &clusterInfo, nil
+}
+
+func evaluateGeneralHealth(cluster *ClusterInfo) GeneralStatus {
+	nodeCount := len(cluster.nodes)
+	generalHealth := "red"
+	if nodeCount > 3 {
+		generalHealth = "green"
+	} else if nodeCount > 1 {
+		generalHealth = "yellow"
+	}
+	return GeneralStatus{
+		Health: generalHealth,
+	}
+}
+
 // Search searches for the cluster information.
 // It try to find cluster information in CR and fallback to storage service when nothing is found.
 func (s *Service) Search(ctx context.Context, request Request) (*Response, error) {
-	response, err := s.searchCR(ctx, request)
+	var err error
+	var cluster *ClusterInfo
+
+	switch s.provider {
+	case "aws":
+		cluster, err = s.searchAWSInfo(ctx, request.ClusterID)
```

Comment:
I think it would be good to handle error in each case to have better stack traces.

Reply:
Done.